test/libevmasm/Optimiser.cpp

```diff
@@ -1160,6 +1160,71 @@ BOOST_AUTO_TEST_CASE(cse_sub_zero)
 	});
 }
 
+BOOST_AUTO_TEST_CASE(cse_remove_redundant_shift_masking)
+{
+	if (!dev::test::Options::get().evmVersion().hasBitwiseShifting())
+		return;
+
+	for (int i = 1; i < 256; i++)
+	{
+		checkCSE({
+				u256(boost::multiprecision::pow(u256(2), i)-1),
+				Instruction::CALLVALUE,
+				u256(256-i),
```

Comment:
Do you also have tests where the mask and the shift amount do not directly match?
I.e. `(x >> (255*8)) & 0xffffffff`?

Reply:
Yes, the for loop following this one does the "not matching" test

Reply:
I did not mean the "not matching" part, but the "over matching". Can you add a test for the one above? Can also just be a yul optimizer test, that is easier to write, I think.

Reply:
No, it can't be a yul test because we can't use SHR in those :)

Reply:
I had a test like that I think and the optimizer just replaced the SHL with `0` because it shifted more than 255

Reply:
Ah sorry, I meant `(x >> (31*8)) & 0xffffffff`

Reply:
Your example will match the pattern and cause an optimization. 
But as far as I was able to validate that is also alright (or was that what you expected the test to result in when you say "overmatching"?)

```
; (2^248) >> (31*8)
        1
; (2^248) >> (31*8) & 0xffffffff
        1
; (2^256) >> (31*8) & 0xffffffff
        256
; (2^256) >> (31*8)
        256
;
```

Reply:
yes, it was supposed to optimize